serializers/avro/src/main/java/io/pravega/schemaregistry/serializer/avro/impl/AvroSerializerFactory.java

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.serializer.avro.impl;
+
+import com.google.common.base.Preconditions;
+import io.pravega.client.stream.Serializer;
+import io.pravega.schemaregistry.serializer.avro.schemas.AvroSchema;
+import io.pravega.schemaregistry.client.SchemaRegistryClient;
+import io.pravega.schemaregistry.common.Either;
+import io.pravega.schemaregistry.contract.data.EncodingInfo;
+import io.pravega.schemaregistry.serializer.shared.impl.AbstractDeserializer;
+import io.pravega.schemaregistry.serializer.shared.impl.AbstractSerializer;
+import io.pravega.schemaregistry.serializer.shared.impl.EncodingCache;
+import io.pravega.schemaregistry.serializer.shared.impl.MultiplexedAndGenericDeserializer;
+import io.pravega.schemaregistry.serializer.shared.impl.MultiplexedDeserializer;
+import io.pravega.schemaregistry.serializer.shared.impl.MultiplexedSerializer;
+import io.pravega.schemaregistry.serializer.shared.impl.SerializerConfig;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.generic.GenericRecord;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static io.pravega.schemaregistry.serializer.shared.impl.SerializerFactoryHelper.initForDeserializer;
+import static io.pravega.schemaregistry.serializer.shared.impl.SerializerFactoryHelper.initForSerializer;
+
+/**
+ * Internal Factory class for Avro serializers and deserializers. 
+ */
+@Slf4j
+public class AvroSerializerFactory {
+    /**
+     * Creates a typed avro serializer for the Schema. The serializer implementation returned from this method is
+     * responsible for interacting with schema registry service and ensures that only valid registered schema can be used.
+     *
+     * Note: the returned serializer only implements {@link Serializer#serialize(Object)}.
+     * It does not implement {@link Serializer#deserialize(ByteBuffer)}.
+     *
+     * @param config     Serializer Config used for instantiating a new serializer.
+     * @param schema Schema container that encapsulates an AvroSchema
+     * @param <T>        Type of event. It accepts either POJO or Avro generated classes and serializes them.
+     * @return A Serializer Implementation that can be used in {@link io.pravega.client.stream.EventStreamWriter} or
+     * {@link io.pravega.client.stream.TransactionalEventStreamWriter}.
+     */
+    public static <T> Serializer<T> serializer(SerializerConfig config, AvroSchema<T> schema) {
+        Preconditions.checkNotNull(config);
+        Preconditions.checkNotNull(schema);
```

Comment:
`@NonNull` Lombok annotation for arguments would eliminate the need for lines 55 and 56, and reduce the code. Same for other methods below too. That being said, it's really a matter of preference.

Reply:
thanks. I will let this be as is for now.